torch/nn/parallel/distributed.py

```diff
@@ -222,7 +243,85 @@ def _register_grad_hooks(self):
                     grad_acc.register_hook(self._make_param_hook(p, device_idx))
                     self._grad_accs.append(grad_acc)
 
+    def _register_nccl_grad_hook(self):
+        """
+        This function registers the callback all-reduction function for the
+        NCCL backend. All gradients will be all reduced in one single step.
+        The NCCL reduction will directly be enqueued into the
+        default CUDA stream. Therefore, no synchronization is needed.
+        """
+        # Creating a new group
+        self.nccl_reduction_group_id = dist.new_group()
+
+        def reduction_fn_nccl():
+            # This function only needs to be called once
+            if not self.need_reduction:
+                return
+            self.need_reduction = False
+            all_grads = [[] for _ in range(len(self._module_copies))]
+            all_grads_coalesced = []
+            # The gradient bucket for the first device of self.device_ids
+            mst_dev_grads_buckets = []
+
+            # Coalesce all the gradients
+            for dev_idx, module in enumerate(self._module_copies):
+                for param in module.parameters():
+                    if not param.requires_grad or param.grad is None:
+                        continue
+                    if param.grad.requires_grad:
+                        raise RuntimeError("DistributedDataParallel only works "
+                                           "with gradients that don't require "
+                                           "grad")
+                    # Adding the gradients for reduction
+                    all_grads[dev_idx].append(param.grad.data)
+
+                # Now bucketing the parameters
+                dev_grads_buckets = _take_tensors(all_grads[dev_idx],
+                                                  self.nccl_reduce_bucket_size)
+                if dev_idx == 0:
+                    mst_dev_grads_buckets = list(dev_grads_buckets)
+                    all_grads_coalesced = \
+                        [[] for _ in range(len(mst_dev_grads_buckets))]
+
+                for bkt_idx, dev_grads in enumerate(dev_grads_buckets):
+                    # Coalescing the bucket
+                    dev_id = self.device_ids[dev_idx]
+                    with torch.cuda.device(dev_id):
+                        dev_grads_coalesced = _flatten_dense_tensors(dev_grads)
+                        all_grads_coalesced[bkt_idx].append(dev_grads_coalesced)
```

Comment:
Well, flattening them all at one time, before you even start broadcasting defeats the whole purpose of bucketing, since you will still use 2x the memory needed for parameters.

Reply:
We are flattening one bucket at one time here. Why do you think it's all at one time?

Reply:
You are flattening them one at a time, but you are first flattening everything and storing the buffers. Basically you're doing this:
```python
flattened = [flatten(bucket) for bucket in buckets] # NB: 2x memory used now
for flat in flattened:
    ... # process flat tensors
```
instead of this:
```python
for bucket in buckets:
    flat = flatten(bucket) # NOTE: flattening inside the loop - is freed at next iteration
    ... # process flat tensor
```